src/charm.py

```diff
@@ -216,6 +250,46 @@ def _on_cluster_relation_changed(self, event: EventBase) -> None:  # noqa: C901
         if not self.state.unit_server.started:
             self.init_server()
 
+        # create services if we expose the charm, no op if not
+        self.update_external_services()
+
+        # since the next steps will 1. update the unit status to active and 2. update the clients,
+        # we want to check if the external access is all good before proceeding
+        if not self.state.endpoints:
+            self.disconnect_clients()
+            event.defer()
+            return
+
+        # if we were already using tls while a network change comes up, we need to expire
+        # existing certificates
+        current_sans = self.tls_manager.get_current_sans()
+
+        current_sans_ip = set(current_sans.sans_ip) if current_sans else set()
+        expected_sans_ip = set(self.tls_manager.build_sans().sans_ip) if current_sans else set()
+        sans_ip_changed = current_sans_ip ^ expected_sans_ip
+
+        current_sans_dns = set(current_sans.sans_dns) if current_sans else set()
+        expected_sans_dns = set(self.tls_manager.build_sans().sans_dns) if current_sans else set()
+        sans_dns_changed = current_sans_dns ^ expected_sans_dns
+
+        if sans_ip_changed or sans_dns_changed:
+            logger.info(
+                (
+                    f'SERVER {self.unit.name.split("/")[1]} updating certificate SANs - '
+                    f"OLD SANs = {current_sans_ip - expected_sans_ip}, "
+                    f"NEW SANs = {expected_sans_ip - current_sans_ip}"
+                )
```

Comment:
**todo:** I think you also want to log the `OLD SANs IP` and `OLD SANs DNS` separately.
There was also a bug with that, see https://github.com/canonical/kafka-operator/pull/276/files for how I fixed it there

Reply:
Thanks, done in 9109bbaa945f026eb29554369eece3d76c5e0318